Apiview readme (#2562)
Azure/azure-sdk-tools: src/dotnet/APIView/APIViewWeb/README.md

# APIView

APIView tool is used by archboard reviewers to review API signatures of all public APIs available in Azure SDK packages. This tool generates public API surface level review which shows all publicly available classes, methods, properties etc. This makes it easier to identify if there are any breaking changes. Currently APIView tool supports following languages:

- C#
- C
- C++
- Java
- JS/TS
- Python
- Go
- Swift

## Why do we need APIView

APIView tool allows to see only stub version of classes, methods, properties, method signatures available in each Azure SDK. This helps architects to review an Azure SDK and helps to identify any potential change that can impact consumers of an Azure SDK. APIView is also used as an enforcing tool to make sure we release a GA version of Azure SDK package only after it is approved by an architect. APIView is also utlized to identify any change at API level in a pull request.

## How to create an API review manually

> API review can be created by uploading an artifact to APIView tool. Type of the artifact is different across each language. Some language, for e.g., Swift requires developer to run parser tool locally to generate stub file and upload json stub file instead of any artifact. Following are the detailed instruction on hwo to create review for each language. 

Reviewer: ```suggestion
API review can be created by uploading an artifact to APIView tool. Type of the artifact is different across each language. Some language, for e.g., Swift requires developer to run parser tool locally to generate stub file and upload json stub file instead of any artifact. Following are the detailed instruction on two to create review for each language. 

```

### C#

> Run `dotnet pack` for the required package to generate nuget file. Upload the resulting .nupkg file using `Create Review` link in APIView.

Reviewer: ```suggestion
Run `dotnet pack` for the required package to generate Nuget file. Upload the resulting .nupkg file using `Create Review` link in APIView.

```

### JS/TS
1. Use api-extractor to generate a [docModel](https://api-extractor.com/pages/setup/generating_docs/) file
2. Upload generated api.json file using `Create Review` link.

### Python
1. Generate wheel for the package. python setup.py bdist_wheel -d [dest_folder]
2. Upload generated whl file

## How does it retrieve public API information

Developers can upload package or an abstract file generated based on each language into API View tool. APIView tool has a language processor for each language it supports and these individual language processor extracts API stub information and create json APIView file that is processed by tool. API review tool stores original uploaded file and generated json file in its data store. Review pages are rendered using this generated json file which contains tokens to present language keywords, links etc. Tool also allows user to recreate review json from stored original file if language package processor itself is updated.

## Types of API reviews

APIView tool shows three different types of reviews based on how it is generated.
- Manual
- Automatic
- Pull request reviews

### Manual

> Manual reviews are created by developers vy uploading artifact as per teh instructions given above. This will allow developers to review API changes if API review is not available from PR branch.

Reviewer: ```suggestion
Manual reviews are created by developers by uploading artifact as per the instructions given above. This will allow developers to review API changes if API review is not available from PR branch.

```

### Automatic

> Azure devops pipeline creates a review or addd a new revision based on the code changes in main branch if package has a new API changes. APIView tool maintains only one review for each package per language. It creates a new revision each time a new change is made at API level for that package. These reviews cannot be deleted or updated with new revisions manually , in other words, this is a locked version of API reviews. Only actions that are allowed on master review is Add/update/remove/Resolve comment and Approve API reviews.

Reviewer: ```suggestion
Azure DevOps pipeline creates a review or add a new revision based on the code changes in main branch if package has a new API change. APIView tool maintains only one review for each package per language. It creates a new revision each time a new change is made at API level for that package. These reviews cannot be deleted or updated with new revisions manually, in other words, this is a locked version of API reviews. Only actions that are allowed on master review is Add/update/remove/Resolve comment and Approve API reviews.

```

As part of build and release pipelines, we will enforce API approval by architect or deputy architect if package version is GA which means we need to ensure latest revision of automatically created review is approved in API review tool to release a GA version package. This is applicable for any package ( both data plane and management plane) that is listed as an artifact in CI yaml.

Automatic API reviews are not listed by default when you login to API review tool. You can view automatic reviews by clicking on "Automatic" button in top right corner in the main page.

### API reviews from pull requests

> PR pipline for Java, C#, JS and Python sends a request to APIView tool to identify if there are any changes made at API level as part of the PR it is processing. APIView tool compared the stub file from PR pipeline against API review revision created from main branch and creates a new review if there is any change. APIView will also add a comment to GitHub PR with a link to API review if there is any change detected. APIView will not crete any review for this PR if it does not have any API level change.

Reviewer: ```suggestion
PR pipeline for Java, C#, JS and Python sends a request to APIView tool to identify if there are any changes made at API level as part of the PR it is processing. APIView tool compared the stub file from PR pipeline against API review revision created from main branch and creates a new review if there is any change. APIView will also add a comment to GitHub PR with a link to API review if there is any change detected. APIView will not create any review for this PR if it does not have any API level change.

```
